kernel/sync_block/src/condvar.rs

```diff
@@ -0,0 +1,107 @@
+use crate::MutexGuard;
+use cpu_local_preemption::hold_preemption;
+use mpmc_queue::Queue;
+use sync::DeadlockPrevention;
+use sync_spin::Spin;
+use task::{get_my_current_task, TaskRef};
+
+/// A condition variable.
+// TODO: Is there even a point to exposing this generic?
+pub struct Condvar<P = Spin>
+where
+    P: DeadlockPrevention,
+{
+    inner: Queue<TaskRef, P>,
+}
+
+impl<P> Condvar<P>
+where
+    P: DeadlockPrevention,
+{
+    pub const fn new() -> Self {
+        Self {
+            inner: Queue::new(),
+        }
+    }
+
+    /// Blocks the current thread until this condition variable receives a
+    /// notification.
+    ///
+    /// Note that this function is susceptible to spurious wakeups. Condition
+    /// variables normally have a boolean predicate associated with them, and
+    /// the predicate must always be checked each time this function returns to
+    /// protect against spurious wakeups.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
```

Comment:
I think we can relax the implicit Sized bound on T here, as Mutexes allow for that too.
```suggestion
    pub fn wait<'a, T: ?Sized>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
```

Reply:
Depends on #962.

Reply:
merged in #962, thanks.